### PR TITLE
[XChange] Get the list of supported currency pairs

### PR DESCRIPTION
**Pull request by tdiesler**
Resolves CAMEL-12167

### CODE REVIEW

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMarketDataProducer.java [hunk 1]
```diff
@@ -45,17 +40,15 @@ public XChangeEndpoint getEndpoint() {
...
+    public void process(Exchange exchange) throws Exception {
+
+        XChangeEndpoint endpoint = getEndpoint();
+        String method = endpoint.getConfiguration().getMethod();
         
-        return 1;
+        if (METHOD_TICKER.equals(method)) {
+            CurrencyPair pair = exchange.getIn().getHeader(HEADER_CURRENCY_PAIR, CurrencyPair.class);
+            Ticker ticker = marketService.getTicker(pair);
+            exchange.getOut().setBody(ticker);
```
Comment (davsclaus) [inline components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMarketDataProducer.java]: Dont set on OUT like that as it losses headers and whatnot during routing.  See this FAQ: http://camel.apache.org/using-getin-or-getout-methods-on-exchange.html

And there is a new getMessage() you can use btw

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        super.doStart();
+
+        XChangeEndpoint endpoint = getEndpoint();
+        String method = endpoint.getConfiguration().getMethod();
+        
+        if (METHOD_CURRENCIES.equals(method)) {
```
Comment (davsclaus) [inline components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java]: This looks really wrong as this consumer is run only once. Before it was scheduled so it calls that once every X period. Why did you change this? If its not scheduled then the 3rd party library must have some kind of event behavior or scheduling on its own that triggers and calls this consumer to process the message in a Camel route

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataProducer.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    @Override
+    public void process(Exchange exchange) throws Exception {
+
+        XChangeEndpoint endpoint = getEndpoint();
+        String method = endpoint.getConfiguration().getMethod();
+        
+        if (METHOD_CURRENCY_METADATA.equals(method)) {
+            Currency curr = exchange.getIn().getHeader(HEADER_CURRENCY, Currency.class);
+            Object body = endpoint.getCurrencyMetaData(curr);
+            exchange.getOut().setBody(body);
```
Comment (davsclaus) [inline components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataProducer.java]: See previous about IN vs OUT

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java (hunk 2 above)
Comment (davsclaus) [reply components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java]: Also the routing should never process in the start method. Those are for starting up resources and whatnot. Only after its started it should poll/trigger on events and route messages - not before.

Comment (tdiesler) [reply components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java]: The [ticker](https://github.com/tdiesler/camel/blob/CAMEL-12167/components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMarketDataProducer.java#L48) functionality is now an ordinary producer. The user ask for the current price of a currency pair when needed (i.e. poll) instead of the component doing these polls on a scheduled basis.

Comment (tdiesler) [reply components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java]: The [metadata consumer](https://github.com/tdiesler/camel/blob/CAMEL-12167/components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java) is indeed something I'm not at all sure about. Here I want to give access to metadata associated with the underlying XChange (not the Camel Exchange). This data becomes available when the endpoint is created and does not change during its lifetime. The low level API is [on the endpoint](https://github.com/tdiesler/camel/blob/CAMEL-12167/components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeEndpoint.java#L95).

Comment (tdiesler) [reply components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMetaDataConsumer.java]: Please see how the [test case](https://github.com/tdiesler/camel/blob/CAMEL-12167/components/camel-xchange/src/test/java/org/apache/camel/component/xchange/MetaDataConsumerTest.java) currently uses the metadata service

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMarketDataProducer.java (hunk 1 above)
Comment (tdiesler) [reply components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeMarketDataProducer.java]: Ok, got it.

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeEndpoint.java [hunk 4]
```diff
@@ -16,15 +16,27 @@
...
+import org.knowm.xchange.currency.Currency;
+import org.knowm.xchange.currency.CurrencyPair;
+import org.knowm.xchange.dto.meta.CurrencyMetaData;
+import org.knowm.xchange.dto.meta.CurrencyPairMetaData;
+import org.knowm.xchange.dto.meta.ExchangeMetaData;
+import org.knowm.xchange.utils.Assert;
 
-@UriEndpoint(firstVersion = "2.21.0", scheme = "xchange", title = "XChange", syntax = "xchange:name", consumerClass = XChangeConsumer.class, label = "api")
-public class XChangeEndpoint extends DefaultPollingEndpoint {
+@UriEndpoint(firstVersion = "2.21.0", scheme = "xchange", title = "XChange", syntax = "xchange:name", label = "api")
```
Comment (davsclaus) [inline components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeEndpoint.java]: You need to set consumerOnly = true, also label = "api" should be label = "currency" or "crypto-currency" or something that is a good label for this kind of components (think label = group)

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeConfiguration.java [hunk 5]
```diff
@@ -17,23 +17,59 @@
...
+
+    static List<String> supportedMethods = Arrays.asList(
+            METHOD_CURRENCIES, METHOD_CURRENCY_METADATA, METHOD_CURRENCY_PAIRS, METHOD_CURRENCY_PAIR_METADATA, METHOD_TICKER);
+    
+    static final String HEADER_CURRENCY = "Currency";
+    static final String HEADER_CURRENCY_PAIR = "CurrencyPair";
+    
     @UriPath(description = "The exchange to connect to") @Metadata(required = "true")
     private String name;
+    @UriParam(description = "The service to call") @Metadata(required = "true")
```
Comment (davsclaus) [inline components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeConfiguration.java]: You can add enum="a,b,c" with the valid values these can have, then its part of the meta-data and tooling would be able to know - we do this for other components.

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeConfiguration.java [hunk 6]
```diff
@@ -46,19 +82,57 @@ public void setName(String name) {
         this.name = name;
     }
 
+    public String getService() {
+        return service;
+    }
+
+    public void setService(String service) {
+        Assert.isTrue(supportedServices.contains(service), "Unsupported service, we have: " + supportedServices);
```
Comment (davsclaus) [inline components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeConfiguration.java]: We tend to do this kind of validation in the doStart method of the component or when creating the producer / consumer (if specific to them) and there is a `ObjectHelper.notNull / notEmpty` we tend to use for this kind.

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeEndpoint.java (hunk 4 above)
Comment (tdiesler) [reply components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeEndpoint.java]: How about blockchain?

File: components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeConfiguration.java (hunk 6 above)
Comment (tdiesler) [reply components/camel-xchange/src/main/java/org/apache/camel/component/xchange/XChangeConfiguration.java]: Yes, I saw the ObjectHelper in Camel. The XChange library however also provides Assert.isTrue() which I couldn't find in Camel

**Comment by tdiesler** [comment]
This is ready for review again
